File: AirHockeyLightClient/AirHockeyLightClient/Noyau/Arbre/Visiteurs/VisiteurDuplication.swift
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurDuplication.swift
/// @author Pierre To
/// @date 2017-10-24
/// @version 0.1
///
/// @addtogroup log3900 LOG3900
/// @{
///////////////////////////////////////////////////////////////////////////////

import GLKit

///////////////////////////////////////////////////////////////////////////
/// @class VisiteurDuplication
/// @brief Permet de dupliquer des noeuds
///
///
/// @author Pierre To
/// @date 2017-10-24
///////////////////////////////////////////////////////////////////////////
class VisiteurDuplication: VisiteurAbstrait {
    
    /// Nombre d'items à dupliquer
    private var nbItems: Int = 0
    
    /// Vecteur trois dimensions du point centre pour la duplication
    private var centreDuplication = GLKVector3()
    
    /// Premier noeud sélectionné pour la duplication
    private var premierNoeud: NoeudPortail?
    
    /// Constructeur
    init() {
        self.nbItems = 0
        self.centreDuplication = GLKVector3.init(v: (0.0, 0.0, 0.0))
    }
    
    /// Cette fonction permet de chercher le centre de la duplication
    func obtenirCentreDuplication() -> GLKVector3 {
        self.centreDuplication = GLKVector3DivideScalar(self.centreDuplication, Float(self.nbItems))
        return self.centreDuplication
    }

    /// Visiter un accéléateur pour la duplication
    func visiterAccelerateur(noeud: NoeudAccelerateur) {
        if noeud.estSelectionne() {
            let arbre = FacadeModele.instance.obtenirArbreRendu()
            
            // Création du noeud
            let noeudDouble = arbre.creerNoeud(typeNouveauNoeud: arbre.NOM_ACCELERATEUR) as! NoeudAccelerateur
            
            // Assigner les mêmes propriétés
            self.copyProperties(node: noeud, nodeCopy: noeudDouble)
            
            // Ajout du noeud à l'arbre de rendu
            let table = arbre.childNode(withName: arbre.NOM_TABLE, recursively: true) as! NoeudTable
            table.addChildNode(noeudDouble)
        }
    }
    
    /// Visiter un maillet pour la duplication
    //virtual void visiterMaillet(NoeudMaillet* noeud);
    
    /// Visiter une table pour la duplication
    func visiterTable(noeud: NoeudTable) {
        // Ne fait rien
    }
    
    /// Visiter un point de contrôle pour la duplication
    func visiterPointControl(noeud: NoeudPointControl) {
        // Ne fait rien
    }
    
    /// Visiter un mur pour la duplication
    func visiterMur(noeud: NoeudMur) {
        if noeud.estSelectionne() {
            let arbre = FacadeModele.instance.obtenirArbreRendu()
            
            // Création du noeud
            let noeudDouble = arbre.creerNoeud(typeNouveauNoeud: arbre.NOM_MUR) as! NoeudMur
            
            // Assigner les mêmes propriétés
            self.copyProperties(node: noeud, nodeCopy: noeudDouble)
            
            // Ajout du noeud à l'arbre de rendu
            let table = arbre.childNode(withName: arbre.NOM_TABLE, recursively: true) as! NoeudTable
            table.addChildNode(noeudDouble)
        }
    }
    
    /// Visiter un portail pour la duplication
    func visiterPortail(noeud: NoeudPortail) {
        if noeud.estSelectionne() && noeud.obtenirOppose().estSelectionne() {
            let arbre = FacadeModele.instance.obtenirArbreRendu()
            
            // Création du noeud
            let noeudDouble = arbre.creerNoeud(typeNouveauNoeud: arbre.NOM_PORTAIL) as! NoeudPortail
            
            // État de l'opération
            if (self.premierNoeud == nil) {
                premierNoeud = noeudDouble
                noeudDouble.assignerOppose(portail: nil)
            }
            else {
                // Relier les deux portails
                noeudDouble.assignerOppose(portail: self.premierNoeud!)
                self.premierNoeud?.assignerOppose(portail: noeudDouble)
                
                // Assigner les mêmes propriétés
                self.copyProperties(node: noeud, nodeCopy: noeudDouble)
                self.copyProperties(node: noeud.obtenirOppose(), nodeCopy: noeudDouble.obtenirOppose())
                
                // Ajouter les portails sur la table
                let table = arbre.childNode(withName: arbre.NOM_TABLE, recursively: true) as! NoeudTable
                table.addChildNode(noeudDouble)
                table.addChildNode(noeudDouble.obtenirOppose())
                
                self.premierNoeud = nil
            }
        }
        else {
            // Pour dupliquer un portail, il faut avoir sélectionner sa paire aussi
            noeud.assignerSelection(selectionne: false)
        }
    }
    
    //virtual void visiterRondelle(NoeudRondelle* noeud);
    
    /// Cette fonction permet de copier les propriétés de l'objet dupliqué
    /// Met aussi le noeud fantôme
    private func copyProperties(node: NoeudCommun, nodeCopy: NoeudCommun) {
        nodeCopy.transform = node.transform
        
        // Déselectionner l'objet à dupliquer
        node.assignerSelection(selectionne: false)
        
        // Sélectionner l'objet dupliqué
        nodeCopy.assignerSelection(selectionne: true)
    
        self.centreDuplication = GLKVector3Add(self.centreDuplication, nodeCopy.obtenirPositionRelative())
        self.nbItems += 1
    }
    
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////

File: Livrable final/Sources/DLL/Arbre/Visiteurs/VisiteurDuplication.cpp
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurDuplication.cpp
/// @author Clint Phoeuk
/// @date 2016-09-28
/// @version 0.1
///
/// @addtogroup inf2990 INF2990
/// @{
///////////////////////////////////////////////////////////////////////////////

#include "VisiteurDuplication.h"
#include "FacadeModele.h"
#include "ModeleEtatJeu.h"
#include "ModeleEtatCreerBoost.h"
#include "ModeleEtatCreerMuret.h"
#include "ModeleEtatCreerPortail.h"

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDuplication::VisiteurDuplication()
///
/// Constructeur pour VisiteurDuplication
///
///	@param[in] Aucun
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurDuplication::VisiteurDuplication(bool sendToServer, WallCreationCallback wallCallback,BoostCreationCallback boostCallback, PortalCreationCallback portalCallback)
	: nbItems_(0), centreDuplication_(glm::vec3(0,0,0)),sendToServer_(sendToServer)
{
	wallCreationCallback_ = wallCallback;
	portalCreationCallback_ = portalCallback;
	boostCreationCallback_ = boostCallback;

	stamp_ = std::vector<NoeudAbstrait*>();
	premierNoeudPortailTrouve_ = false;
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDuplication::VisiteurDuplication()
///
/// Destructeur pour VisiteurDuplication
///
///	@param[in] Aucun
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurDuplication::~VisiteurDuplication() {

}


////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDuplication::visiterAccelerateur(NoeudAccelerateur* noeud)
///
/// Cette fonction permet de visiter un accélérateur pour la duplication.
///
///	@param[in] noeud : Noeud de l'objet 
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurDuplication::visiterAccelerateur(NoeudAccelerateur* noeud) {
	
	if (noeud->estSelectionne()) {
		// Création du noeud
		NoeudAbstrait* noeudDouble = FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->creerNoeud(ArbreRenduINF2990::NOM_ACCELERATEUR);

		// Assigner les mêmes propriétés
		copyProperties(noeud, noeudDouble);
		noeud->assignerSelection(false);

		// Ajout du noeud à l'arbre de rendu
		FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->ajouter(noeudDouble);

		stamp_.push_back(noeudDouble);


		
		if (sendToServer_)
		{
			boostCreationCallback_(noeud->getUUID(), glm::value_ptr(noeud->obtenirPositionRelative()), noeud->obtenirRotation().y, glm::value_ptr(noeud->obtenirScale()));
		}

	}
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDuplication::visiterMur(NoeudMur* noeud)
///
/// Cette fonction permet de visiter un mur pour la duplication.
///
///	@param[in] noeud : Noeud de l'objet 
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurDuplication::visiterMur(NoeudMur* noeud) {
	
	if (noeud->estSelectionne()) {
		// Création du noeud
		NoeudAbstrait* noeudDouble = FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->creerNoeud(ArbreRenduINF2990::NOM_MUR);

		// Assigner les mêmes propriétés
		copyProperties(noeud, noeudDouble);

		noeud->assignerSelection(false);

		stamp_.push_back(noeudDouble);
		if (sendToServer_)
		{
			wallCreationCallback_(noeud->getUUID(), glm::value_ptr(noeud->obtenirPositionRelative()), noeud->obtenirRotation().y, glm::value_ptr(noeud->obtenirScale()));
		}

		// Ajout du noeud à l'arbre de rendu
		FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->ajouter(noeudDouble);
	}
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDuplication::visiterPortail(NoeudPortail* noeud)
///
/// Cette fonction permet de visiter un portail pour la duplication.
///
///	@param[in] noeud : Noeud de l'objet 
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurDuplication::visiterPortail(NoeudPortail* noeud) {
	if (noeud->estSelectionne() && noeud->obtenirOppose()->estSelectionne()) {

		// Création du noeud
		NoeudPortail* noeudDouble = (NoeudPortail*)FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->creerNoeud(ArbreRenduINF2990::NOM_PORTAIL);

		// Assigner les mêmes propriétés
		copyProperties(noeud, noeudDouble);

		stamp_.push_back(noeudDouble);


		// État de l'opération	
		if (!premierNoeudPortailTrouve_) {
			premierNoeud_ = noeudDouble;

			premierNoeudAEnvoye_ = noeud;

			noeudDouble->assignerOppose(nullptr);
			premierNoeudPortailTrouve_ = true;

		}

		else {
			// Relier les deux portails
			noeudDouble->assignerOppose(premierNoeud_);
			premierNoeud_->assignerOppose(noeudDouble);


			premierNoeudAEnvoye_->assignerSelection(false);
			noeud->assignerSelection(false);


			if (sendToServer_)
			{
				portalCreationCallback_(premierNoeudAEnvoye_->getUUID(), glm::value_ptr(premierNoeudAEnvoye_->obtenirPositionRelative()), (premierNoeudAEnvoye_->obtenirRotation().y), glm::value_ptr(premierNoeudAEnvoye_->obtenirScale()),
					noeud->getUUID(), glm::value_ptr(noeud->obtenirPositionRelative()), noeud->obtenirRotation().y, glm::value_ptr(noeud->obtenirScale()));
			}
			premierNoeudPortailTrouve_ = false;
		}

		// Ajout du noeud à l'arbre de rendu
		FacadeModele::obtenirInstance()->obtenirArbreRenduINF2990()->ajouter(noeudDouble);
	}
	else
		noeud->assignerSelection(false);
}

void VisiteurDuplication::visiterRondelle(NoeudRondelle * noeud)
{
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDuplication::obtenirCentreDuplication()
///
/// Cette fonction permet de chercher le centre de la duplication.
///
///	@param[in] Aucun
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
glm::vec3 VisiteurDuplication::obtenirCentreDuplication() {
	centreDuplication_ /= nbItems_;
	return centreDuplication_;
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDuplication::copyProperties(NoeudAbstrait* node, NoeudAbstrait* nodeCopy)
///
/// Cette fonction permet de copier les propriétés de l'objet dupliqué.
///
///	@param[in] noeud : Noeud de l'objet à copier
///	@param[in] nodeCopy : Copie du noeud de l'objet 
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurDuplication::copyProperties(NoeudAbstrait* node, NoeudAbstrait* nodeCopy) {
	nodeCopy->assignerPositionRelative(node->obtenirPositionRelative());
	nodeCopy->scale(node->obtenirScale());
	nodeCopy->rotate(node->obtenirRotation().y, glm::vec3(0, 1, 0));

	centreDuplication_ += nodeCopy->obtenirPositionRelative();
	nbItems_++;
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////